Data race in `TestWatcherCompacted`

## Flaky Test

### Which jobs are failing
```
==================
WARNING: DATA RACE
Read at 0x00c0002e8960 by goroutine 2380:
  github.com/tikv/pd/server/config.(*Config).Clone()
      /home/runner/work/pd/pd/server/config/config.go:481 +0x131
  github.com/tikv/pd/server.(*Server).GetConfig()
      /home/runner/work/pd/pd/server/server.go:942 +0x87
  github.com/tikv/pd/tests.(*TestServer).GetConfig()
      /home/runner/work/pd/pd/tests/cluster.go:182 +0x14d
  github.com/tikv/pd/tests/server/watch_test.TestWatcherCompacted.func2()
      /home/runner/work/pd/pd/tests/server/watch/leader_watch_test.go:98 +0x6c
  github.com/stretchr/testify/assert.Eventually.func1()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:1902 +0x33

Previous write at 0x00c0002e8960 by goroutine 420:
  github.com/tikv/pd/pkg/utils/typeutil.(*Duration).UnmarshalJSON()
      /home/runner/work/pd/pd/pkg/utils/typeutil/duration.go:50 +0x2a4
  encoding/json.(*decodeState).literalStore()
      /opt/hostedtoolcache/go/1.23.8/x64/src/encoding/json/decode.go:854 +0x394a
  encoding/json.(*decodeState).value()
      /opt/hostedtoolcache/go/1.23.8/x64/src/encoding/json/decode.go:388 +0x224
  encoding/json.(*decodeState).object()
      /opt/hostedtoolcache/go/1.23.8/x64/src/encoding/json/decode.go:755 +0x1453
  encoding/json.(*decodeState).value()
      /opt/hostedtoolcache/go/1.23.8/x64/src/encoding/json/decode.go:374 +0xae
  encoding/json.(*decodeState).unmarshal()
      /opt/hostedtoolcache/go/1.23.8/x64/src/encoding/json/decode.go:181 +0x28f
  encoding/json.Unmarshal()
      /opt/hostedtoolcache/go/1.23.8/x64/src/encoding/json/decode.go:108 +0x22b
  github.com/tikv/pd/pkg/mcs/resourcemanager/server.(*Manager).Init()
      /home/runner/work/pd/pd/pkg/mcs/resourcemanager/server/manager.go:127 +0x36a
  github.com/tikv/pd/pkg/mcs/resourcemanager/server.(*Manager).Init-fm()
      <autogenerated>:1 +0x47
  github.com/tikv/pd/server.(*Server).campaignLeader()
      /home/runner/work/pd/pd/server/server.go:1709 +0x128e
  github.com/tikv/pd/server.(*Server).leaderLoop()
      /home/runner/work/pd/pd/server/server.go:1659 +0x1afc
  github.com/tikv/pd/server.(*Server).startServerLoop.gowrap1()
      /home/runner/work/pd/pd/server/server.go:643 +0x33

Goroutine 2380 (running) created at:
  github.com/stretchr/testify/assert.Eventually()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:1902 +0x3d5
  github.com/stretchr/testify/require.Eventually()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.9.0/require/require.go:398 +0xd1
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x91
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x4f8
  fmt.Fscanf()
      /opt/hostedtoolcache/go/1.23.8/x64/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /opt/hostedtoolcache/go/1.23.8/x64/src/fmt/scan.go:114 +0x264
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:657 +0x18e
  fmt.(*ss).doScanf()
      /opt/hostedtoolcache/go/1.23.8/x64/src/fmt/scan.go:1230 +0x43b
  fmt.Fscanf()
      /opt/hostedtoolcache/go/1.23.8/x64/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /opt/hostedtoolcache/go/1.23.8/x64/src/fmt/scan.go:114 +0x184
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x91
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x4f8
  fmt.Fscanf()
      /opt/hostedtoolcache/go/1.23.8/x64/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /opt/hostedtoolcache/go/1.23.8/x64/src/fmt/scan.go:114 +0x184
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x91
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x4f8
  github.com/syndtr/goleveldb/leveldb.(*DB).recoverJournal()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:476 +0xa6
  github.com/syndtr/goleveldb/leveldb.openDB()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:125 +0x86d
  github.com/syndtr/goleveldb/leveldb.Open()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:197 +0x37c
  fmt.Fscanf()
      /opt/hostedtoolcache/go/1.23.8/x64/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /opt/hostedtoolcache/go/1.23.8/x64/src/fmt/scan.go:114 +0x184
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x91
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x4f8
  github.com/syndtr/goleveldb/leveldb.(*session).recover.func1()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/session.go:134 +0xbd
  runtime.deferreturn()
      /opt/hostedtoolcache/go/1.23.8/x64/src/runtime/panic.go:611 +0x5d
  github.com/syndtr/goleveldb/leveldb.Open()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:183 +0x11b
  github.com/syndtr/goleveldb/leveldb.OpenFile()
      /home/runner/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:219 +0x7d
  github.com/tikv/pd/pkg/storage/kv.NewLevelDBKV()
      /home/runner/work/pd/pd/pkg/storage/kv/levedb_kv.go:36 +0x9e
  github.com/tikv/pd/pkg/storage.newLevelDBBackend()
      /home/runner/work/pd/pd/pkg/storage/leveldb_backend.go:64 +0xbb
  github.com/tikv/pd/pkg/storage.NewRegionStorageWithLevelDBBackend()
      /home/runner/work/pd/pd/pkg/storage/storage.go:70 +0xb5
  github.com/tikv/pd/server.(*Server).startServer()
      /home/runner/work/pd/pd/server/server.go:460 +0x10a5
  github.com/tikv/pd/server.(*Server).Run()
      /home/runner/work/pd/pd/server/server.go:606 +0x1ed
  github.com/tikv/pd/tests.(*TestServer).Run()
      /home/runner/work/pd/pd/tests/cluster.go:123 +0x1d5
  github.com/tikv/pd/tests.RunServer.func1()
      /home/runner/work/pd/pd/tests/cluster.go:576 +0x44
==================
    testing.go:1399: race detected during execution of test
--- FAIL: TestWatcherCompacted (23.85s)
```


### CI link

https://github.com/tikv/pd/actions/runs/14439066442/job/40490629124?pr=9184

### Reason for failure (if possible)


### Anything else

--- a/tests/server/watch/leader_watch_test.go
+++ b/tests/server/watch/leader_watch_test.go
@@ -95,6 +95,6 @@ func TestWatcherCompacted(t *testing.T) {
 	err = pd2.Run()
 	re.NoError(err)
 	testutil.Eventually(re, func() bool {
-		return pd2.GetLeader().GetName() == pd1.GetConfig().Name
+		return cluster.WaitLeader() != "" && pd2.GetLeader().GetName() == pd1.GetConfig().Name
 	})
 }

--- a/tests/server/watch/leader_watch_test.go
+++ b/tests/server/watch/leader_watch_test.go
@@ -95,6 +95,6 @@ func TestWatcherCompacted(t *testing.T) {
 	err = pd2.Run()
 	re.NoError(err)
 	testutil.Eventually(re, func() bool {
-		return cluster.WaitLeader() != "" && pd2.GetLeader().GetName() == pd1.GetConfig().Name
+		return len(cluster.WaitLeader()) > 0 && pd2.GetLeader().GetName() == pd1.GetConfig().Name
 	})
 }

